miden-lib/asm/kernels/transaction/api.masm

```diff
@@ -161,32 +172,36 @@ end
 
 #! Sets an item in the account storage. Panics if the index is out of bounds.
 #!
-#! Stack: [index, V', 0, 0, 0]
-#! Output: [R', V]
+#! Stack: [index, NEW_VALUE]
+#! Output: [OLD_VALUE]
 #!
 #! - index is the index of the item to set.
-#! - V' is the value to set.
-#! - V is the previous value of the item.
-#! - R' is the new storage root.
-export.set_account_item
-    # if the transaction is being executed against a faucet account then assert
-    # index != FAUCET_STORAGE_DATA_SLOT (reserved slot)
-    dup exec.account::get_faucet_storage_data_slot eq
-    exec.account::get_id exec.account::is_faucet
-    and assertz.err=ERR_FAUCET_RESERVED_DATA_SLOT
-    # => [index, V', 0, 0, 0]
-
+#! - NEW_VALUE is the value to set.
+#! - OLD_VALUE is the previous value of the item.
+export.set_account_item.2
     # authenticate that the procedure invocation originates from the account context
     exec.authenticate_account_origin
-    # => [index, V', 0, 0, 0]
+    # => [index, NEW_VALUE]
+
+    push.0 drop                                             # TODO: remove line, see miden-vm/#1122
+    emit.ACCOUNT_STORAGE_BEFORE_SET_ITEM_EVENT
+    # => [index, NEW_VALUE]
+
+    # store index and NEW_VALUE for later to emit event. Note: loc_store pops the Felt value
+    loc_store.0 loc_storew.1 loc_load.0
+    # => [index, NEW_VALUE]
 
     # set the account storage item
     exec.account::set_item
-    # => [R', V, 0, 0, 0]
+    # => [OLD_VALUE]
+
+    # emit event to signal that an account storage item is being updated
+    padw loc_loadw.1 loc_load.0 emit.ACCOUNT_STORAGE_AFTER_SET_ITEM_EVENT drop dropw
+    # => [OLD_VALUE]
```

Comment:
nit: I would probably move `drop dropw` to line 203 below and update the comments accordingly.